Resync config.toml and suggest better settings (#795)
rust-lang/rustc-dev-guide: src/building/how-to-build-and-run.md

```toml
[llvm]
# Enables LLVM assertions, which will check that the LLVM bitcode generated
# by the compiler is internally consistent. These are particularly helpful
# if you edit `codegen`.
# Indicates whether the LLVM assertions are enabled or not
assertions = true

[rust]
# This will make your build more parallel; it costs a bit of runtime

> # performance perhaps (less inlining) but it's worth it.

Reviewer: Was this from `config.toml.example`? It looks like these two lines explained why the `= 0` value was set, and this was lost.

Reviewer 2: Yes, this is from `config.toml.example`. I'll add it back in https://github.com/rust-lang/rust/pull/76141.

#       enabled becomes *unusably slow* (e.g. rust-lang/rust#24840
#       reported a 25x slowdown) and bootstrapping the supposed
#       "maximally debuggable" environment (notably libstd) takes
#       hours to build.
#
debug = true

# Number of codegen units to use for each compiler invocation. A value of 0
# means "the number of cores on this machine", and 1+ is passed through to the
# compiler.
codegen-units = 0

# Debuginfo level for most of Rust code, corresponds to the `-C debuginfo=N` option of `rustc`.
# `0` - no debug info
# `1` - line tables only - sufficient to generate backtraces that include line
#       information and inlined functions, set breakpoints at source code
#       locations, and step through execution in a debugger.
# `2` - full debug info with variable and type information
# Can be overridden for specific subsets of Rust code (rustc, std or tools).
# Debuginfo for tests run with compiletest is not controlled by this option
# and needs to be enabled separately with `debuginfo-level-tests`.
#
# Defaults to 2 if debug is true
debuginfo-level = 1

# Whether to always use incremental compilation when building rustc

> incremental = true

Reviewer: According [to Cargo documentation](https://doc.rust-lang.org/nightly/cargo/reference/profiles.html#codegen-units), the incremental default for `codegen-units` is 256, which makes `codegen-units = 0` unnecessary/undesirable (unless you're compiling Rust on a machine with more than 256 hardware cores/threads, heh).

So my understanding is that when adding `incremental = true`, we also need to remove `codegen-units = 0`. cc @Mark-Simulacrum

# Emits extra output from tests so test failures are debuggable just from logfiles.

> verbose-tests = true

Reviewer: Why was this added? The default is that failures are listed at the end, without noisy output while running the tests.

```toml
[llvm]
# Indicates whether the LLVM assertions are enabled or not

> assertions = true

Reviewer 2: Why do we recommend this? I don't expect most new contributors to be editing codegen and it will slow down the compiler.

Author: :+1: I think you're right here, I've just left it there because it was there already but given that we are fixing suggestions it may be a good idea to remove it.

What others think about this? why was this originally added?.

Author: Discussed here https://rust-lang.zulipchat.com/#narrow/stream/196385-t-compiler.2Fwg-rustc-dev-guide/topic/meeting.2007.2E14.2E2020

#       enabled becomes *unusably slow* (e.g. rust-lang/rust#24840
#       reported a 25x slowdown) and bootstrapping the supposed
#       "maximally debuggable" environment (notably libstd) takes
#       hours to build.
#
debug = true

# Number of codegen units to use for each compiler invocation. A value of 0
# means "the number of cores on this machine", and 1+ is passed through to the
# compiler.
codegen-units = 0

# Debuginfo level for most of Rust code, corresponds to the `-C debuginfo=N` option of `rustc`.
# `0` - no debug info
# `1` - line tables only - sufficient to generate backtraces that include line
#       information and inlined functions, set breakpoints at source code
#       locations, and step through execution in a debugger.
# `2` - full debug info with variable and type information
# Can be overridden for specific subsets of Rust code (rustc, std or tools).
# Debuginfo for tests run with compiletest is not controlled by this option
# and needs to be enabled separately with `debuginfo-level-tests`.
#
# Defaults to 2 if debug is true
debuginfo-level = 1

# Whether to always use incremental compilation when building rustc
incremental = true

> # Emits extraneous output from tests to ensure that failures of the test
> # harness are debuggable just from logfiles.
> verbose-tests = true

Reviewer 2: ```suggestion
# Emits extra output from tests so test failures are debuggable just from logfiles.
verbose-tests = true
```
Extraneous means it's not useful or irrelevant.

Author: @jyn514 :+1: but consider changing this on rust-lang/rust, I've just copied what it's there already.

Reviewer: This change doesn't seem right, specifically "test harness" refers to *the code running the tests*, rather than the tests themselves. And it is extraneous for the common case of e.g. `libtest` not crashing.
